src/lint/linter.py

```diff
@@ -0,0 +1,172 @@
+   doing typechecking on them - that's done at runtime anyway more accurately than
+   we can here. This is somewhat unnecessary but again may help some workflows
+   where the linter might hint you not to commit with obvious problems.
+ - Detection of duplicates in argument lists - most usually this is useful for deps,
+   but it applies to anything since there's no reason to have a duplicate in any
+   argument list to any builtin Please function.
+
+Lint warnings can be suppressed on a per-line basis by adding a trailing comment
+saying either `# nolint` or `# lint:disable=iterkeys-used`.
+At present there is no way to disable it by scope or completely for a file.
+"""
+
+import argparse
+import ast
+import json
+import pkg_resources
+import re
+import sys
+
+
+SYNTAX_ERROR = 'syntax-error'
+ITERITEMS_USED = 'iteritems-used'
+ITERVALUES_USED = 'itervalues-used'
+ITERKEYS_USED = 'iterkeys-used'
+UNSORTED_SET_ITERATION = 'unsorted-set-iteration'
+UNSORTED_DICT_ITERATION = 'unsorted-dict-iteration'
+NON_KEYWORD_CALL = 'non-keyword-call'
+DEPRECATED_FUNCTION = 'deprecated-function'
+INCORRECT_ARGUMENT = 'incorrect-argument'
+UNNECESSARY_DUPLICATE = 'unnecessary-duplicate'
+
+
+ERROR_DESCRIPTIONS = {
+    SYNTAX_ERROR: 'Syntax error',
+    ITERITEMS_USED: 'dict.iteritems called, use dict.items instead',
+    ITERVALUES_USED: 'dict.itervalues called, use dict.values instead',
+    ITERKEYS_USED: 'dict.iterkeys called, use dict.keys instead (or just iterate the dict)',
+    UNSORTED_SET_ITERATION: 'Iteration of sets is not ordered, use sorted()',
+    UNSORTED_DICT_ITERATION: 'Iteration of dicts is not ordered, use sorted()',
+    NON_KEYWORD_CALL: 'Call to builtin rule without using keyword arguments',
+    DEPRECATED_FUNCTION: 'The function include_defs is deprecated, use subinclude() instead',
+    INCORRECT_ARGUMENT: 'Unknown argument to built-in function',
+    UNNECESSARY_DUPLICATE: 'Unnecessary duplicate in argument',
+}
+
+BANNED_ATTRS = {
+    'iteritems': ITERITEMS_USED,
+    'itervalues': ITERVALUES_USED,
+    'iterkeys': ITERKEYS_USED,
+}
+
+UNSORTED_CALLS = {
+    'set': UNSORTED_SET_ITERATION,
+    'dict': UNSORTED_DICT_ITERATION,
+}
+
+JSON = json.loads(pkg_resources.resource_string('src.parse', 'rule_args.json'))
+WHITELISTED_FUNCTIONS = {'subinclude', 'glob', 'include_defs', 'licenses'}
+BUILTIN_FUNCTIONS = {k: v for k, v in JSON['functions'].items() if k not in WHITELISTED_FUNCTIONS}
+DEPRECATED_FUNCTIONS = {'include_defs'}
+
+
+def is_suppressed(code, line):
+    """Returns True if the given code is suppressed on this line."""
+    if '#' not in line:
+        return False
+    comment = line[line.index('#') + 1:]
+    return 'nolint' in comment or re.search('lint: *disable=' + code, comment)
+
+
+def _lint(contents):
+    try:
+        tree = ast.parse(contents)
+    except SyntaxError as err:
+        yield err.lineno, SYNTAX_ERROR
+        return
+
+    for n in ast.walk(tree):
```

Comment:
Definitely keep the fixture files -- they are effectively an acceptance test -- and have the lint checks inside the same module. My thinking was just splitting them in separate functions. Something along the lines of the following:

```
def _check_unsorted_iteration(n):
   if isinstance(n, ast.For):
       if isinstance(n.iter, ast.Call) and isinstance(n.iter.func, ast.Name):
           if n.iter.func.id in UNSORTED_CALLS:
               return n.lineno, UNSORTED_CALLS[n.iter.func.id]
       elif isinstance(n.iter, (ast.Dict, ast.DictComp)):
           return n.lineno, UNSORTED_DICT_ITERATION
       elif isinstance(n.iter, (ast.Set, ast.SetComp)):
           return n.lineno, UNSORTED_SET_ITERATION
  return None

def _check_builtin_args(n):
   ...

def _check_deprecated_functions(n):
  ...

...
LINTER_FNS = [...]
def _lint(contents):
  ...
  lint_errors = (fn(n) for n in ast.walk(tree) for fn in LINTER_FNS)  # or just spell out the functions explicitly
  return (e for e in lint_errors if e)  # avoiding using filter vs itertools.ifilter :P
```